### ISSUE TITLE
Build an Online Incremental Learning Layer Between Full Model Retrains

### ISSUE THREAD

**Issue by Inkman007** [Ledger-Lenz/Ledgerlens-core#315]
## Summary

`cli.py retrain-check` is the only mechanism LedgerLens has for adapting the RF/XGBoost/LightGBM ensemble to new patterns — it's a full batch retrain triggered by PSI drift (`detection/drift_monitor.py`), gated by an AUC-ROC promote/revert check, and it runs on a schedule (README recommends daily via cron/systemd). Between retrains, confirmed analyst feedback (`detection/feedback_store.py`'s `AnalystFeedbackStore`) currently only feeds `detection/adaptive_reweighter.py`'s Thompson-sampling *ensemble weight* adjustment and `detection/ensemble_reweighter.py`'s Bayesian model-averaging weights — neither actually updates any model's learned parameters. This issue adds an online/incremental learning layer that applies small, bounded gradient-style nudges to the base models from confirmed feedback between full retrains, with automatic reversion if online drift exceeds a safety bound.

## Background & Context

README's "Continuous Retraining" section documents the current adaptation loop precisely: PSI is computed per feature, drift is declared when ≥3 features exceed `PSI_THRESHOLD` (0.20, `detection/drift_monitor.py`), a full retrain runs on the *original synthetic training distribution*, and the new models are promoted only if `AUC-ROC ≥ previous version` — otherwise `cli.py retrain-check` reverts automatically. This is a coarse, slow-cycle mechanism: it discards everything learned since the last retrain and starts from the same synthetic baseline every time, and it only runs when explicitly scheduled (daily, per the documented cron/systemd examples). Meanwhile `detection/feedback_store.py`'s `AnalystFeedbackStore` already captures ground-truth corrections continuously via `POST /v1/feedback` (README's "Analyst Feedback Endpoints" table), and `detection/drift_monitor.py`'s `PerformanceMonitor.check_degradation()` already computes rolling F1 against a `baseline_f1` and raises `ModelDegradationAlert` when it drops more than `f1_threshold_drop` — but today that alert only *triggers* the next scheduled full retrain; it has no faster-acting remediation available.

`detection/adaptive_reweighter.py`'s `ThompsonSamplingReweighter` is the closest existing analog to online learning in this codebase: it maintains Beta-distribution posteriors per classifier and updates `alphas`/`betas` incrementally from `reward` signals, persisted to a `bandit_state` SQLite table and refreshed every `_UPDATE_INTERVAL_SECONDS` (15 minutes). But it only reweights the *combination* of three frozen models — it cannot correct a systematic bias any individual model has learned, because none of RF/XGBoost/LightGBM's own parameters ever change outside a full `cli.py retrain-check` run.

The gap: there is no bridge between "we have 50 new confirmed analyst labels since the last retrain" and "the model incorporates them" short of waiting for the next scheduled PSI-triggered or forced retrain. An online incremental layer — using River (a Python library purpose-built for online ML) or `partial_fit`-style incremental gradient boosting updates — closes that gap with small, bounded parameter nudges, while keeping the existing full-retrain pipeline as the ground-truth periodic reset (so online drift cannot compound indefinitely between full retrains).

## Objectives

- [ ] Implement `OnlineIncrementalLayer` in `detection/online_learning.py` wrapping a lightweight online model (River `HoeffdingTreeClassifier` or an incremental logistic-regression correction head) trained on confirmed feedback from `AnalystFeedbackStore` between full retrains
- [ ] Define the composition rule: the online layer does **not** replace RF/XGBoost/LightGBM's predictions — it produces a bounded *correction delta* (`Δp ∈ [-ONLINE_MAX_DELTA, +ONLINE_MAX_DELTA]`, default `±0.15` probability) applied on top of the frozen ensemble's output, analogous to how `RiskScore.combine` already composes multiple bounded adjustments (`sandwich_weight`, `pdc_discount_weight`, `benford_copula_weight`)
- [ ] Wire `POST /v1/feedback` (the existing analyst feedback endpoint) to also call `OnlineIncrementalLayer.learn_one(features, ground_truth)` immediately upon a confirmed correction, not just write to `AnalystFeedbackStore`
- [ ] Implement a safety-bound monitor: track the online layer's own prediction drift (e.g. rolling mean absolute correction magnitude) and automatically **reset** the online layer to a no-op (zero correction) state if its corrections exceed `ONLINE_DRIFT_SAFETY_BOUND` — reusing `detection/adaptive_reweighter.py`'s existing `cusum_detect()` CUSUM test for the shift-detection logic rather than inventing a new one
- [ ] Persist online layer state (River model serialization, or incremental model weights) to SQLite following the existing `bandit_state`/`rolling_window_checkpoints` checkpoint pattern, so it survives restarts
- [ ] Reset the online layer to empty/no-op state every time `cli.py retrain-check` promotes new base models — the online layer's job is to bridge *to* the next full retrain, not accumulate indefinitely across retrain boundaries
- [ ] Expose online layer status via `GET /admin/online-learning-status` (admin-key gated, consistent with `/admin/drift-reports`): sample count since last reset, current safety-bound distance, last correction magnitude distribution

## Technical Requirements

**Online layer** (`detection/online_learning.py`, new module):

```python
class OnlineIncrementalLayer:
    """Bounded online correction applied between full batch retrains.

    Wraps a River HoeffdingTreeClassifier (or incremental logistic regression)
    trained continuously on AnalystFeedbackStore corrections. Never replaces
    the frozen RF/XGBoost/LightGBM ensemble output — only nudges it within
    ONLINE_MAX_DELTA.
    """

    def __init__(self, max_delta: float = 0.15, safety_bound: float = 0.30): ...

    def learn_one(self, features: dict[str, float], ground_truth: int) -> None:
        """Incorporate one confirmed analyst correction."""

    def correction(self, features: dict[str, float]) -> float:
        """Return Δp ∈ [-max_delta, +max_delta] to add to the base ensemble probability."""

    def check_safety_bound(self) -> bool:
        """Runs cusum_detect() (reused from adaptive_reweighter.py) over the
        recent correction-magnitude series. Returns True if the bound was
        breached and the layer was reset to no-op.
        """

    def reset(self, reason: str) -> None:
        """Called on retrain promotion or safety-bound breach."""
```

**Composition into `RiskScore.combine`** — a new optional parameter mirroring the existing bounded-adjustment pattern:

```python
score = RiskScore.combine(
    wallet=wallet, asset_pair=asset_pair,
    benford_mad=..., benford_mad_threshold=..., ml_probability=ml_probability,
    ml_confidence=..., 
    online_correction=online_layer.correction(features),   # NEW: Δp, default 0.0
    ...
)
```

`RiskScore.combine` applies `ml_probability_adjusted = clip(ml_probability + online_correction, 0.0, 1.0)` before computing `ml_component`, with `online_correction=0.0` reproducing exactly the current, unmodified behaviour.

**Safety bound check** — reuses `detection/adaptive_reweighter.py`'s `cusum_detect(errors, threshold, slack)` directly:

```python
recent_corrections = online_layer.get_recent_correction_magnitudes(window=200)
if cusum_detect(recent_corrections, threshold=ONLINE_CUSUM_THRESHOLD, slack=ONLINE_CUSUM_SLACK):
    online_layer.reset(reason="cusum_safety_bound_breach")
    logger.warning("OnlineIncrementalLayer reset: correction drift exceeded safety bound")
```

**Configuration** (`.env.example`):

```
ONLINE_LEARNING_ENABLED=false
ONLINE_MAX_DELTA=0.15
ONLINE_DRIFT_SAFETY_BOUND=0.30
ONLINE_CUSUM_THRESHOLD=5.0
ONLINE_CUSUM_SLACK=0.1
ONLINE_MIN_SAMPLES_BEFORE_ACTIVE=20   # correction() returns 0.0 until this many learn_one() calls
```

## Security Considerations

- The online layer trains on `AnalystFeedbackStore` data, which is admin-key gated at ingestion (`POST /v1/feedback` per README already requires the admin key) — no new attack surface is introduced for *submitting* feedback, but this issue raises the stakes of that endpoint's authorization, since a compromised admin key could now be used to poison live scoring, not just historical feedback records. Call this out explicitly in the security review for this feature
- `ONLINE_MAX_DELTA` is a hard cap enforced in code (`clip`), not merely a training hyperparameter — even a fully poisoned online layer cannot move any single score by more than `ONLINE_MAX_DELTA * 100` points, bounding the blast radius of any feedback-poisoning attack
- The CUSUM safety-bound reset must fire automatically and log at `WARNING`/alert-worthy severity — this is the primary defense against slow-drift poisoning (an adversary submitting many small "confirmations" that each individually looks legitimate but cumulatively bias the online layer)
- Online layer state persisted to SQLite must not be loaded via pickle (River supports safe serialization via its own format or `cloudpickle` with integrity checking) — follow the same "no raw pickle for untrusted-adjacent state" discipline `rolling_window.py`'s security notes already establish for checkpoint JSON

## Testing Requirements

- `tests/test_online_learning.py`
- Test: `OnlineIncrementalLayer.correction()` returns `0.0` before `ONLINE_MIN_SAMPLES_BEFORE_ACTIVE` feedback samples have been learned
- Test: `correction()` is always clipped to `[-ONLINE_MAX_DELTA, +ONLINE_MAX_DELTA]` regardless of how extreme the underlying online model's raw prediction is
- Test: `RiskScore.combine(..., online_correction=0.0)` produces byte-identical output to the current `RiskScore.combine` signature (backward compatibility)
- Test: repeated large corrections trigger `check_safety_bound()` → `reset()`, and `correction()` returns `0.0` immediately after reset
- Test: `cli.py retrain-check` promotion path calls `OnlineIncrementalLayer.reset(reason="retrain_promoted")`
- Test: state persistence round-trip (save, restart process, reload) preserves learned corrections
- Test: `POST /v1/feedback` triggers `learn_one()` exactly once per confirmed correction (no double-counting on retry)
- Test: `GET /admin/online-learning-status` returns `503` without admin key, correct status payload with it

## Documentation Requirements

- New `docs/online_learning.md`: motivation (the gap between confirmed feedback and the next scheduled full retrain), architecture (bounded correction layer, not a replacement model), safety-bound mechanism, interaction with `cli.py retrain-check`'s promote/revert cycle
- Update README.md's "Continuous Retraining" section with a subsection on the online incremental layer, cross-referencing `docs/online_learning.md`
- Update `.env.example` with the new `ONLINE_*` variables

## Definition of Done

- [ ] All objectives completed
- [ ] Tests pass (`pytest`)
- [ ] No regressions on existing test suite
- [ ] PR reviewed and approved

## For Contributors

**When applying for this issue, please specify:**
- Your area of specialty (e.g., online/streaming ML, MLOps, continuous learning systems)
- Relevant experience with: River or similar online ML libraries, incremental gradient boosting (`partial_fit` patterns), safety-bounded model updates, drift/poisoning detection (CUSUM or similar)
- Your approach or initial thoughts on the implementation
- Estimated time to complete

**Ideal contributor profile:** ML engineer with production experience running online/incremental learning safely — someone who has designed bounded-correction or shadow-model patterns before and takes model-poisoning risk seriously, since this issue explicitly widens the blast radius of the existing analyst-feedback trust boundary.

**Comment by Inkman007** [comment]
Closing as part of a full issue-tracker reset: the project's open issues have been replaced with a new curated set of 125 advanced-level issues covering the codebase end to end. No action needed here; if this item is still relevant, please open a fresh issue referencing it.
